x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/ibmwatsonx/IbmWatsonxRequest.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.external.request.ibmwatsonx;
+
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.entity.ByteArrayEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.message.BasicHeader;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.xcontent.XContentParseException;
+import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xpack.core.common.socket.SocketAccess;
+import org.elasticsearch.xpack.inference.external.request.Request;
+import org.elasticsearch.xpack.inference.services.settings.DefaultSecretSettings;
+
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+
+public interface IbmWatsonxRequest extends Request {
+
+    static void decorateWithBearerToken(HttpPost httpPost, DefaultSecretSettings secretSettings) {
+        final Logger logger = LogManager.getLogger(IbmWatsonxRequest.class);
+        String bearerTokenGenUrl = "https://iam.cloud.ibm.com/identity/token";
+        String bearerToken = "";
+
+        try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
+            HttpPost httpPostForBearerToken = new HttpPost(bearerTokenGenUrl);
+
+            String body = "grant_type=urn:ibm:params:oauth:grant-type:apikey&apikey=" + secretSettings.apiKey().toString();
+            ByteArrayEntity byteEntity = new ByteArrayEntity(body.getBytes(StandardCharsets.UTF_8));
+
+            httpPostForBearerToken.setEntity(byteEntity);
+            httpPostForBearerToken.setHeader(HttpHeaders.CONTENT_TYPE, "application/x-www-form-urlencoded");
+
+            bearerToken = SocketAccess.doPrivileged(() -> {
+                HttpResponse response = httpClient.execute(httpPostForBearerToken);
```

Comment:
@prwhelan

I tried to [add response validation](https://github.com/elastic/elasticsearch/pull/111770/commits/f3cd615b8eee1f39536175dee7fd4588a691f319) based on the status code, similar to error handling in the [IbmWatsonxResponseHandler](https://github.com/elastic/elasticsearch/pull/111770/files#diff-107f5e3d773721875e1240bb966dcbd38d2cee7b844b80355393236e881c2e86R50), where exceptions are thrown for specific status codes. However, this change requires appending `throws IOException` to createHttpRequest(), which affects several test files related to other providers.

To avoid affecting unrelated tests, can we address this validation in a separate PR since this one is specific to IBM Watsonx? I’ll create a GitHub ticket to track this and include the validation in a future PR. WDYT?

Reply:
I'm still worried that a user will get an unactionable and misleading error from an authentication issue that they can resolve themselves if they get the appropriate error message.

Why not combine parts of what you have in both commits?
https://github.com/prwhelan/elasticsearch/commit/b1e67eb0ead09f8a781c5666e9e54cd110ee4da7